### PR TITLE
Touch the tablet with your finger

### PR DESCRIPTION
**Pull request by hyperlogic**
By default the user can use the index finger of their avatar to touch buttons on the tablet.  If their avatar is missing a IndexFinger4 joint, or they have disabled this via General Preferences, the stylus is used instead.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87b61fdb7fed7df312bab5479518e18197d8dd76 available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-87b61fd.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-87b61fd.dmg

**Comment by PhilipRosedale** [comment]
Feedback:  Works great, I like it better than the stylus.  

I think the one thing we need here is to curl the fingers under when near the tablet so that the index finger first contacts the tablet surface, rather than the outstretched fingers we have now.

**Comment by hifi-gustavo** [comment]
Build console output for commit: fffcb4c1a80fa74fda62b1d03acb17d4ee97479d available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-fffcb4c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-fffcb4c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f41845710042c935d078ef0d04111ca99480053b available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-f418457.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-f418457.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d142c3d69bfd851b57f6901bf24440b3c0ec4697 available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-d142c3d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-d142c3d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 465a9e2008f18e893b27e33e87e4c58c5c475bd2 available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-465a9e2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-465a9e2.dmg

**Comment by hyperlogic** [comment]
QA Test Plan:
* Ensure that both fingers can be used to press buttons on the tablet.  If your avatar has HandIndex4 joints.  The default blue avatar should have the required joints.
* Test choosing avatars that do not have this joint, they should revert to using the stylus.
* Test using the "Prefer Avatar Finger Over Stylus" General Preference.  If not set, the stylus should always appear, when the hand is near the tablet.  Even for avatars with HandIndex4 Joints.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4f03c06a948ce37283f07247e73ffc793fa24f2e available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-4f03c06.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-4f03c06.dmg

File: libraries/script-engine/src/Vec3.h [hunk 1]
```diff
@@ -37,6 +37,15 @@
  * @property {float} z Z-coordinate of the vector.
  */
 
+/**jsdoc
+ * A 4-dimensional vector.
```
Comment (sethalves) [inline libraries/script-engine/src/Vec3.h]: 3-dimensional?  no w?

File: scripts/system/controllers/handControllerGrab.js [hunk 2]
```diff
@@ -28,7 +28,7 @@ Script.include("/~/system/libraries/controllers.js");
 //
 
 var WANT_DEBUG = false;
-var WANT_DEBUG_STATE = false;
+var WANT_DEBUG_STATE = true;
```
Comment (sethalves) [inline scripts/system/controllers/handControllerGrab.js]: turn this back off, please

File: libraries/script-engine/src/Vec3.h (hunk 1 above)
Comment (hyperlogic) [inline libraries/script-engine/src/Vec3.h]: There wasn't anywhere more reasonable to place the jsdoc for vec2 and vec4 objects, as there is no Vec2.h or Vec4.h.  So, I just put them here.

File: scripts/system/controllers/handControllerGrab.js (hunk 2 above)
Comment (hyperlogic) [inline scripts/system/controllers/handControllerGrab.js]: Good catch

**Comment by hifi-gustavo** [comment]
Build console output for commit: 644e29a43d63bf8029e4f71bf4d6d576d693eaab available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-644e29a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-644e29a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bc256f3e8f01ce07b5740f00099e7dac59c1b4ef available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-bc256f3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-bc256f3.dmg

**Comment by Nex-Pro** [comment]
testing this

**Comment by Nex-Pro** [comment]
Tested this on win10 and vive.

@hyperlogic 

In general this works ok but the tapping action seems a little hit or miss. Less reliable than with the stylus.
Tested with the being of light. And one of the Morph3d avatars.

I then selected the robot and the small skeleton from the marketplace and both really could only use the tablet with the stylii as reaching the tablet was a problem. Neither one automatically got the stylii. I had to select them from the settings.

Is there a specific avatar that has the missing joint. If so a link would help.

**Comment by hifi-gustavo** [comment]
Build console output for commit: e56f02d94f8c499acb50b099830d4dce05be0875 available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-e56f02d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-e56f02d.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 45febe8565371ee961b1db88d20b49c4a2ccffc3 available at: https://gist.github.com/0371aa775ac86b3fc2571c86fe1b55a8<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-45febe8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9754-45febe8.dmg

**Pull request merged**